File: jarvis/packages/shared/src/nep.ts
import { z } from "zod";

// Node Execution Protocol: docs/jarvis/12-stack-and-contracts.md §17.6

export const NEP_PROTOCOL_VERSION = "1.0";

export const NepGrant = z.object({ decision_id: z.string(), signature: z.string(), fingerprint: z.string(), expires_at: z.string() });
export const NepLease = z.object({ lease_id: z.string(), fencing_token: z.number().int() });

export const NepInvoke = z.object({
  invocation_id: z.string(),
  action_id: z.string().optional(),
  capability: z.string(),
  params: z.unknown(),
  grant: NepGrant.optional(),
  lease: NepLease.optional(),
  task_revision: z.number().int(),
  policy_revision: z.number().int(),
  idempotency_key: z.string(),
  deadline: z.string(),
  observation_ref: z.object({ id: z.string(), max_age_ms: z.number() }).optional(),
});
export type NepInvoke = z.infer<typeof NepInvoke>;

export const NepEvent = z.object({
  invocation_id: z.string(), seq: z.number().int(), kind: z.enum(["progress", "observation", "log"]), data: z.unknown(),
});
export type NepEvent = z.infer<typeof NepEvent>;

export const NepCancel = z.object({ invocation_id: z.string(), mode: z.enum(["cooperative", "kill"]) });
export type NepCancel = z.infer<typeof NepCancel>;

// JSON-RPC 2.0 framing shared by every process boundary.
export interface JsonRpcRequest { jsonrpc: "2.0"; id?: string | number; method: string; params?: unknown }
export interface JsonRpcResponse { jsonrpc: "2.0"; id: string | number | null; result?: unknown; error?: { code: number; message: string; data?: unknown } }

File: jarvis/native/Jarvis.Contracts/Nep.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Jarvis.Contracts;

/// <summary>Node Execution Protocol (docs/jarvis/12-stack-and-contracts.md §17.6). Field names match the JSON Schema in jarvis/schemas/nep.</summary>
public static class Nep
{
    public const string ProtocolVersion = "1.0";
}

public sealed record NepGrant(
    [property: JsonPropertyName("decision_id")] string DecisionId,
    [property: JsonPropertyName("signature")] string Signature,
    [property: JsonPropertyName("fingerprint")] string Fingerprint,
    [property: JsonPropertyName("expires_at")] string ExpiresAt);

public sealed record NepLease(
    [property: JsonPropertyName("lease_id")] string LeaseId,
    [property: JsonPropertyName("fencing_token")] long FencingToken);

public sealed class NepInvoke
{
    [JsonPropertyName("invocation_id")] public string InvocationId { get; set; } = "";
    [JsonPropertyName("action_id")] public string? ActionId { get; set; }
    [JsonPropertyName("capability")] public string Capability { get; set; } = "";
    [JsonPropertyName("params")] public JsonElement Params { get; set; }
    [JsonPropertyName("grant")] public NepGrant? Grant { get; set; }
    [JsonPropertyName("lease")] public NepLease? Lease { get; set; }
    [JsonPropertyName("task_revision")] public long TaskRevision { get; set; }
    [JsonPropertyName("policy_revision")] public long PolicyRevision { get; set; }
    [JsonPropertyName("idempotency_key")] public string IdempotencyKey { get; set; } = "";
    [JsonPropertyName("deadline")] public string Deadline { get; set; } = "";
}

/// <summary>Shared error vocabulary (05 §11.6).</summary>
public static class ErrorCodes
{
    public const string InvalidInput = "invalid_input", MissingPermission = "missing_permission", UnsupportedOperation = "unsupported_operation",
        TransientServiceError = "transient_service_error", UncertainExternalEffect = "uncertain_external_effect", Conflict = "conflict",
        Timeout = "timeout", Expired = "expired", Cancelled = "cancelled", InternalError = "internal_error", PreconditionChanged = "precondition_changed",
        ExternalRefusal = "external_refusal";
}

public sealed record StructuredError(
    [property: JsonPropertyName("code")] string Code,
    [property: JsonPropertyName("message")] string Message,
    [property: JsonPropertyName("retryable")] bool Retryable,
    [property: JsonPropertyName("effect_state")] string EffectState,
    [property: JsonPropertyName("source")] ErrorSource Source,
    [property: JsonPropertyName("details")] Dictionary<string, object?>? Details = null);

public sealed record ErrorSource(
    [property: JsonPropertyName("capability_id")] string CapabilityId,
    [property: JsonPropertyName("executor")] string Executor,
    [property: JsonPropertyName("node_id")] string NodeId);

/// <summary>An executor's result; the Coordinator turns it into a ToolResult.</summary>
public sealed class ExecResult
{
    [JsonPropertyName("status")] public string Status { get; set; } = "ok";
    [JsonPropertyName("effect_state")] public string EffectState { get; set; } = "none";
    [JsonPropertyName("output")] public object? Output { get; set; }
    [JsonPropertyName("error")] public StructuredError? Error { get; set; }
    [JsonPropertyName("evidence")] public List<Dictionary<string, object?>> Evidence { get; set; } = new();
}

public sealed class JarvisException(string code, string message, string effectState = "none") : Exception(message)
{
    public string Code { get; } = code;
    public string EffectState { get; } = effectState;
    public StructuredError ToStructured(string capability, string executor) =>
        new(Code, Message, Code is ErrorCodes.TransientServiceError or ErrorCodes.Conflict, EffectState, new ErrorSource(capability, executor, "node_local"));
}
